### ISSUE TITLE
Support high-dimensional BO

### ISSUE THREAD

**Issue by jacktang** [bayesian-optimization/BayesianOptimization#499]
**Is your feature request related to a problem? Please describe.**
Bayesian optimization for high-dimensional constraint problem

**Describe the solution you'd like**
The solution called Sparse Axis-Aligned Subspace BO (SAASBO) is described in the paper: https://arxiv.org/abs/2103.00349

**References or alternative approaches**
* https://arxiv.org/abs/2103.00349

**Are you able and willing to implement this feature yourself and open a pull request?**
- [x] Yes, I would like to try to implement it

**Comment by till-m** [comment]
Hi @jacktang,

I haven't had the time to read this paper, but considering that this is essentially an acquisition strategy, do you think you could implement this simply by [adding a new acquisition function](https://github.com/bayesian-optimization/BayesianOptimization/blob/master/examples/acquisition_functions.ipynb)?

**Comment by jacktang** [comment]
Hello @till-m 

Yes,  I am going to implement it the way you suggested. And I noticed that you had merge the acquisition function branch to master. Great job! 👍

**Comment by MandaKausthubh** [comment]
Hi @jacktang,

Can we use a Python Library, such as numpyro, for this, which simplifies and provides a pre-packaged version of the No U-Turn Sampling required in SAASBO?

**Comment by jacktang** [comment]
Hello @MandaKausthubh, actually I used numpyro for inner saasgp class. now i am stuck in  solving the high dimensional constrained problem, and plan to submit the code in next two or three weeks.

**Comment by jacktang** [comment]
Hello @till-m, SAASBO may bring some changes : 1) sample based GP (SAASGP) rather than analytical GP (standard GP from sklearn) 2) Sample-based acquisition functions.

**Comment by jacktang** [comment]
I organized the HDBO implementation which mentioned in paper "Standard Gaussian Process is All You Need for High-Dimensional Bayesian Optimization" [here](https://github.com/jacktang/HDBO/blob/main/hdbo-torch.py). Check it if you are interested in it. And try to implement the idea by sklearn GP in next weeks.

**Comment by jacktang** [comment]
Hello @till-m , I had interests to learn the paper, and now only implemented the recommended HDBO by gpytorch. It's great you've draft and ready to push :). I'd like to test the implementation especially for the constrained HDBO if needed.

**Comment by jacktang** [comment]
Hello @till-m , I add [Hartmann6 problem for HDBO](https://github.com/jacktang/BayesianOptimization/blob/feature-scaling/examples/hdbo.ipynb), the conclusion is current sklearn GP can't solve the optimization problem. I guess feature scaling and length scale are some part of the key to solving the problem, but may be not the all part(?)

**Comment by till-m** [comment]
Hey @jacktang,

great that you worked on the implementation! I had a quick look at your code and I notice you didn't set the kernel to be anisotropic (what they call "ARD" in the paper). You can do that by adding this line before running `.maximize`

```python
from sklearn.gaussian_process.kernels import Matern
optimizer.set_gp_params(
    kernel=Matern(nu=2.5, length_scale=np.ones(dim) * np.sqrt(dim)),
)
```

NB: This will significantly increase the runtime, and I still didn't get close to the score you mentioned in the markdown cell 🤔
